pkg/cache/inmemory.go

```diff
@@ -0,0 +1,302 @@
+	"github.com/thanos-io/thanos/pkg/model"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	DefaultInMemoryCacheConfig = InMemoryCacheConfig{
+		MaxSize:     250 * 1024 * 1024,
+		MaxItemSize: 125 * 1024 * 1024,
+	}
+)
+
+const (
+	maxInt = int(^uint(0) >> 1)
+	// Size of a golang slice i.e unsafe.Sizeof([]byte{}) on amd64.
+	sliceHeaderSize = 16
+	// As the cache key is of type String its size is 8 bytes.
+	keyHeaderSize = 8
+)
+
+// InMemoryCacheConfig holds the in-memory cache config.
+type InMemoryCacheConfig struct {
+	// MaxSize represents overall maximum number of bytes cache can contain.
+	MaxSize model.Bytes `yaml:"max_size"`
+	// MaxItemSize represents maximum size of single item.
+	MaxItemSize model.Bytes `yaml:"max_item_size"`
+}
+
+type InMemoryCache struct {
+	mtx sync.Mutex
+
+	logger           log.Logger
+	lru              *lru.LRU
+	maxSizeBytes     uint64
+	maxItemSizeBytes uint64
+
+	curSize uint64
+
+	evicted          prometheus.Counter
+	requests         prometheus.Counter
+	hits             prometheus.Counter
+	hitsExpired      prometheus.Counter
+	added            prometheus.Counter
+	current          prometheus.Gauge
+	currentSize      prometheus.Gauge
+	totalCurrentSize prometheus.Gauge
+	overflow         prometheus.Counter
+}
+
+type cacheDataWithTTLWrapper struct {
+	data       []byte
+	expiryTime time.Time
+}
+
+// parseInMemoryCacheConfig unmarshals a buffer into a InMemoryCacheConfig with default values.
+func parseInMemoryCacheConfig(conf []byte) (InMemoryCacheConfig, error) {
+	config := DefaultInMemoryCacheConfig
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return InMemoryCacheConfig{}, err
+	}
+
+	return config, nil
+}
+
+// NewInMemoryCache creates a new thread-safe LRU cache for chunks and metadata and ensures the total cache
+// size approximately does not exceed maxBytes.
+func NewInMemoryCache(name string, logger log.Logger, reg prometheus.Registerer, conf []byte) (*InMemoryCache, error) {
+	config, err := parseInMemoryCacheConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewInMemoryCacheWithConfig(name, logger, reg, config)
+}
+
+// NewInMemoryCacheWithConfig creates a new thread-safe LRU cache chunks and metadata ensures the total cache
+// size approximately does not exceed maxBytes.
+func NewInMemoryCacheWithConfig(name string, logger log.Logger, reg prometheus.Registerer, config InMemoryCacheConfig) (*InMemoryCache, error) {
+	if config.MaxItemSize > config.MaxSize {
+		return nil, errors.Errorf("max item size (%v) cannot be bigger than overall cache size (%v)", config.MaxItemSize, config.MaxSize)
+	}
+
+	c := &InMemoryCache{
+		logger:           logger,
+		maxSizeBytes:     uint64(config.MaxSize),
+		maxItemSizeBytes: uint64(config.MaxItemSize),
+	}
+
+	c.evicted = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_store_cache_items_evicted_total",
+		Help:        "Total number of items that were evicted from the inmemory cache.",
```

Comment:
This help indicates that this metric is _only_ for inmemory. However, this is not the case for the metric name. If you have a look at the metrics defined here for exmaple: https://github.com/thanos-io/thanos/blob/b5e7b2b493783d89933874293e5c6a97a5b444b5/pkg/cache/memcached.go#L35-L40
You see that the name and help relfect the cache type.

Reply:
Same for all the other metrics.

Reply:
Changed the prefix to `thanos_cache_inmemory_`